### PR TITLE
Improves MDB performance with instance pool

### PR DESCRIPTION
**Pull request by otaviojava**
This goal of this proposal is to improve performance in the message-driven bean creation using a pool of 10 objects (default value).
The strategy is to keep these objects live, so they can be reused instead of every time create a new one.

I have observed that the Websphere MQ RAR does not provide pool endpoints, and the MDB container was initially written with the assumption that most RARs do.

### CODE REVIEW

**Comment by brunobat** [comment]
@tveronezi I've took a look at the failing tests:
 ActivationContainerOverwriteIdContainerConfigurationTest.test:101->testConnection:110 null
 ActivationContainerOverwriteBothConfigurationTest.test:102->testConnection:111 null
 ActivationContainerOverwriteReadTheContainerConfigurationTest.test:100->testConnection:109 null
 ActivationContainerOverwriteMessageConfigurationTest.test:101->testConnection:110 null
 ActivationContainerOverwriteSystemContainerIdTest.test:109->testConnection:118 null
 ActivationContainerOverwriteSystemTest.test:109->testConnection:118 null

They all have the same problem. The message is not received by the MDB and the test fails after the timeout.
After dinging on the code, it seems that the property override is not being properly applied after the creation of https://github.com/apache/tomee/pull/117/files#diff-71712cba31f2983a986542ae146e781d

As an example, in ActivationContainerOverwriteBothConfigurationTest, when we hardcode the destination:
```
    @MessageDriven(activationConfig = {
            @ActivationConfigProperty(propertyName = "destinationType", propertyValue = "javax.jms.Queue"),
            @ActivationConfigProperty(propertyName = "destination", propertyValue = "target")
    })
    public static class Listener implements MessageListener {
```
The test passes. 
When we are booting up, on the  EndpointFactory constructor, the destination was never overriden by the property by
```
.p("mdbs.activation.destination", "target")
 ```
and still shown the `toBeOverwrite`

Seems that we are not properly overriding the properties of the MDBs.

File: container/openejb-core/src/main/java/org/apache/openejb/core/mdb/PoolEndpointHandler.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+        // call afterDelivery on the container
+        boolean exceptionThrown = false;
+        try {
+            container.afterDelivery(instance);
+        } catch (final SystemException se) {
+            exceptionThrown = true;
+
+            final Throwable throwable = se.getRootCause() != null ? se.getRootCause() : se;
+            throwable.printStackTrace();
```
Comment (SvetlinZarev) [inline container/openejb-core/src/main/java/org/apache/openejb/core/mdb/PoolEndpointHandler.java]: Why not logging it instead ?

File: container/openejb-core/src/main/java/org/apache/openejb/core/mdb/PoolEndpointHandler.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        state = State.RELEASED;
+
+        // notify the container
+        try {
+            container.release(deployment, instance);
+        } finally {
+            try {
+                instanceManager.poolInstance(new ThreadContext(deployment, null), instance);
+            } catch (OpenEJBException e) {
+                e.printStackTrace();
```
Comment (SvetlinZarev) [inline container/openejb-core/src/main/java/org/apache/openejb/core/mdb/PoolEndpointHandler.java]: Why not logging it instead?

File: container/openejb-core/src/main/java/org/apache/openejb/core/instance/InstanceManager.java [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+
+    public void destroy() {
+        if (executor != null) {
+            executor.shutdown();
+            try {
+                if (!executor.awaitTermination(10000, MILLISECONDS)) {
+                    java.util.logging.Logger.getLogger(this.getClass().getName()).log(Level.WARNING, getClass().getSimpleName() + " pool  timeout expired");
+                }
+            } catch (final InterruptedException e) {
+                Thread.interrupted();
```
Comment (SvetlinZarev) [inline container/openejb-core/src/main/java/org/apache/openejb/core/instance/InstanceManager.java]: The interrupted state should have been cleared before throwing this exception.

File: container/openejb-core/src/main/java/org/apache/openejb/core/mdb/AbstractEndpointHandler.java [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+                } catch (final ApplicationServerInternalException e) {
+                    throwable = throwable == null ? e.getCause() : throwable;
+                } catch (final UnavailableException e) {
+                    throwable = throwable == null ? e : throwable;
+                }
+            }
+        }
+
+        if (throwable != null) {
+            throwable.printStackTrace();
```
[the remark this answers is not in the document]
Comment (tveronezi) [reply container/openejb-core/src/main/java/org/apache/openejb/core/mdb/AbstractEndpointHandler.java]: I will remove this line. We throw the exception just after that anyway.

File: container/openejb-core/src/main/java/org/apache/openejb/core/mdb/AbstractEndpointHandler.java [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+
+        // call afterDelivery on the container
+        boolean exceptionThrown = false;
+        try {
+            container.afterDelivery(instance);
+        } catch (final SystemException se) {
+            exceptionThrown = true;
+
+            final Throwable throwable = se.getRootCause() != null ? se.getRootCause() : se;
+            throwable.printStackTrace();
```
[the remark this answers is not in the document]
Comment (tveronezi) [reply container/openejb-core/src/main/java/org/apache/openejb/core/mdb/AbstractEndpointHandler.java]: ditto

File: container/openejb-core/src/main/java/org/apache/openejb/core/mdb/PoolEndpointHandler.java (hunk 1 above)
Comment (tveronezi) [reply container/openejb-core/src/main/java/org/apache/openejb/core/mdb/PoolEndpointHandler.java]: ditto

File: container/openejb-core/src/main/java/org/apache/openejb/core/mdb/MdbInstanceManager.java [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+        } catch (final ResourceException e) {
+            throw new OpenEJBException(e);
+        }
+
+        final Options options = new Options(beanContext.getProperties());
+        // Finally, fill the pool and start it
+        if (!options.get("BackgroundStartup", false) && min > 0) {
+            final ExecutorService es = Executors.newFixedThreadPool(min);
+            for (int i = 0; i < min; i++) {
+                es.submit(new InstanceCreatorRunnable(maxAge, i, min, maxAgeOffset, data, supplier));
```
[the remark this answers is not in the document]
Comment (tveronezi) [reply container/openejb-core/src/main/java/org/apache/openejb/core/mdb/MdbInstanceManager.java]: The `i` changes for every instance. right?

**Comment by jgallimore** [comment]
@SvetlinZarev I dig some digging on the handling of the InterruptedException, specifically https://stackoverflow.com/questions/3976344/handling-interruptedexception-in-java and chapter 7 of the referenced book, "Java Concurrency in Action".

If the `InterruptedException` is not propagated to its caller, it should restore the interruption status after catching `InterruptedException` with `Thread.currentThread().interrupt();`. `Thread.interrupt()` is a convenience shorthand for `Thread.currentThread().interrupt()`, so I think this looks ok.

Many thanks for taking the time to review this, I'm enjoying watching the collaboration on this. Thiago is helping address feedback, and I'll join in, if there is any left. I'll also go through and review as well.

File: container/openejb-core/src/main/java/org/apache/openejb/core/mdb/MdbInstanceManager.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (jgallimore) [reply container/openejb-core/src/main/java/org/apache/openejb/core/mdb/MdbInstanceManager.java]: It does, and is used in the calculation of the offset in the runnable, so this probably needs to stay as it is.

**Comment by SvetlinZarev** [comment]
@jgallimore I've seen this pattern and that was the reason I was asking. In the code it's Thread.interrupted(), which clears the interruption flag,   instead of Thread.interrupt(), so it looked like a typo/no-op.

**Comment by jgallimore** [comment]
Damn, I missed that. :-) Tired eyes, I guess.

**Pull request merged**